File: 卡码/代码随想录/1/09-动态规划/C++/dp/02-背包问题/src/01-01背包/Kama_0046_1.cpp
//46. 携带研究材料
#include <iostream>
#include <vector>
using namespace std;

int main(){
	int m,n;
	scanf("%d %d",&m,&n);
	int weight[m];
	int value[m];
	for(int i=0;i<m;i++){
		scanf("%d",&weight[i]);
	}
	for(int i=0;i<m;i++){
		scanf("%d",&value[i]);
	}
	vector<int> dp(n+1,0);
	for(int i=0;i<m;i++){
		for(int j=n;j>=weight[i];j--){
			dp[j]=max(dp[j],dp[j-weight[i]]+value[i]);
		}
	}
	printf("%d\n",dp[n]);
  system("pause");
  return 0;
}

File: 卡码/代码随想录/1/09-动态规划/Java/dp/src/main/java/com/kurumi/dp/_02_背包问题/_01_01背包/Kama_0046_1.java
//46. 携带研究材料
package com.kurumi.dp._02_背包问题._01_01背包;

import java.util.Scanner;

public class Kama_0046_1 {
  public static void main(String[] args) {
    Scanner scanner = new Scanner(System.in);
    int m = scanner.nextInt();
    int n = scanner.nextInt();
    int[] weight = new int[m + 1];
    int[] value = new int[m + 1];
    for (int i = 0; i < m; i++) {
      weight[i] = scanner.nextInt();
    }
    for (int i = 0; i < m; i++) {
      value[i] = scanner.nextInt();
    }
    int[] dp = new int[n + 1];
    for (int i = 0; i < m; i++) {
      for (int j = n; j >=weight[i]; j--) {
        dp[j]=Math.max(dp[j], dp[j-weight[i]]+value[i]);
      }
    }
    System.out.println(dp[n]);
  }
}
